content/blogposts/2023/slow-to-simd.md

```diff
@@ -0,0 +1,508 @@
+// instructions The slices must be of equal length and that length must be a
+// multiple of 64.
+TEXT ·DotVNNI(SB), NOSPLIT, $0-52
+	// Offsets based on slice header offsets.
+	// To check, use `GOARCH=amd64 go vet`
+	MOVQ a_base+0(FP), AX
+	MOVQ b_base+24(FP), BX
+	MOVQ a_len+8(FP), DX
+
+    ADDQ AX, DX // end pointer
+
+	// Zero our accumulators
+	VPXORQ Z0, Z0, Z0 // positive
+	VPXORQ Z1, Z1, Z1 // negative
+
+	// Fill Z2 with 128
+	MOVD $0x80808080, R9
+	VPBROADCASTD R9, Z2
+
+blockloop:
+	CMPQ AX, DX
+	JE reduce
+
+	VMOVDQU8 (AX), Z3
+	VMOVDQU8 (BX), Z4
+
+	// The VPDPBUSD instruction calculates of the dot product 4 columns at a
+	// time, accumulating into an i32 vector. The problem is it expects one
+	// vector to be unsigned bytes and one to be signed bytes. To make this
+	// work, we make one of our vectors unsigned by adding 128 to each element.
+	// This causes us to overshoot, so we keep track of the amount we need
+	// to compensate by so we can subtract it from the sum at the end.
+	//
+	// Effectively, we are calculating SUM((Z3 + 128) · Z4) - 128 * SUM(Z4).
+
+	VPADDB Z3, Z2, Z3   // add 128 to Z3, making it unsigned
+	VPDPBUSD Z4, Z3, Z0 // Z0 += Z3 dot Z4
+	VPDPBUSD Z4, Z2, Z1 // Z1 += broadcast(128) dot Z4
+
+	ADDQ $64, AX
+	ADDQ $64, BX
+	JMP blockloop
+
+reduce:
+    // Subtract the overshoot from our calculated dot product
+	VPSUBD Z1, Z0, Z0 // Z0 -= Z1
+
+    // Sum Z0 horizontally. There is no horizontal sum instruction, so instead
+    // we sum the upper and lower halves of Z0, fold it in half again, and
+    // repeat until we are down to 1 element that contains the final sum.
+    VEXTRACTI64X4 $1, Z0, Y1
+    VPADDD Y0, Y1, Y0
+
+	VEXTRACTI128 $1, Y0, X1
+	VPADDD X0, X1, X0
+
+	VPSRLDQ $8, X0, X1
+	VPADDD X0, X1, X0
+
+	VPSRLDQ $4, X0, X1
+	VPADDD X0, X1, X0
+
+	// Store the reduced sum
+	VMOVD X0, R8
+
+end:
+	MOVL R8, ret+48(FP)
+	VZEROALL
+	RET
+```
+
+</details>
+
+This implementation yielded another 21% improvement. Not bad!
+
+<Chart
+    rows={[
+        {name: "DotNaive", value: 941900},
+        {name: "DotUnroll4", value: 1286000},
+        {name: "DotBCE", value: 1407000},
+        {name: "DotInt8BCE", value: 1240000},
+        {name: "DotAVX2", value: 7000000},
+        {name: "DotVNNI", value: 8765000},
+    ]}
+/>
+
+# What's next?
+
+Well, I'm pretty happy with an 9.3x increase in throughput and a 4x reduction in memory usage, so I'll probably leave it
+here.
+
+The real life answer here is probably "use an index". There is a ton of good work out there focused on making nearest
+neighbor search fast, and there are plenty of batteries-included vector DBs that make it pretty easy to deploy. 
+
+_However_, if you want some fun food for thought, a colleague of built a proof-of-concept [dot product on the
+GPU](https://github.com/sourcegraph/sourcegraph/compare/main...nsc/embeddings-fun#diff-eed4a741ebe632c484dce236a3f4b1eee16e9a2bec6749003b3dbc41449c497c).
```

Comment:
This link feels fragile, maybe he could open it as a draft PR ?

Reply:
I've updated the link to use a tag so it should be safe to treat as a permalink